### PR TITLE
Add AMI comparisons to postprocessing

### PR DESCRIPTION
**Pull request by mdahlhausen**
Pull request overview
---------------------

This PR adds AMI data comparisons from EULP to the postprocessing workflow. The AMI data is available on AWS in an Athena table.

This includes:
- methods to download and process timeseries AMI data
- methods to download and process timeseries ComStock data
- adds an `ami_comparison.csv` to create a 10k `buildstock.csv` sample to compare against the AMI regions
- adds plotting for stacked end use plots by season and day type, either by kwh per sf, or annual or daytype normalizations
- adds plotting for load duration curves
- adds a test

### Pull Request Author

This pull request makes changes to (select all the apply):
 - [x] Postprocessing

Author pull request checklist:
 - [x] Tagged the pull request with the appropriate label (documentation, infrastructure, sampling, workflow measure, upgrade measure, reporting measure, postprocessing) to help categorize changes in the release notes.
 - [x] Added tests for postprocessing
 - [x] 10k+ test run
 - [ ] ComStock documentation updated
 - [ ] Changes reflected in `README.md` files
 - [x] Added 'See ComStock License' language to first two lines of each code file
 - [ ] All new and existing tests pass the CI

### Specific Review Requests
- What, if any content should I add to the postprocessing README?
- Should some of AMI comparisons be included in the release or documentation? Perhaps a separate calibration report to accompany the reference documentation?
- I don't often write in python, so my syntax may be messy. code review it please.
- What other plots should I include?

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a code review on GitHub
 - [ ] All related changes have been implemented: data and method additions, changes, tests
 - [ ] If fixing a defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] Reviewed change documentation
 - [ ] Ensured code files contain License reference
 - [ ] Results differences are reasonable
 - [ ] CI status: all tests pass

### CODE REVIEW

**Comment by mdahlhausen** [comment]
There are some messy warning about enumerations when I run the test. Those seem to come from other parts of the code. Perhaps worth making a github issue if you can track it down.

File: postprocessing/compare_comstock_to_ami.py
Comment (eringold) [inline postprocessing/compare_comstock_to_ami.py:1]: Adding ability to run as a script, simliar to other 'compare_comstock_to' scripts

File: postprocessing/comstockpostproc/ami.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        logger.info(f'Downloading {self.dataset_name}')
+        self.download_truth_data()
+        if reload_from_csv:
+            file_name = f'AMI long.csv'
+            file_path = os.path.join(self.output_dir, file_name)
+            if not os.path.exists(file_path):
+                 raise FileNotFoundError(
+                    f'Cannot find {file_path} to reload data, set reload_from_csv=False to create CSV.')
+            logger.info(f'Reloading from CSV: {file_path}')
+            self.ami_timeseries_data = pd.read_csv(file_path, low_memory=False, index_col='timestamp', parse_dates=True)
```
Comment (eringold) [inline postprocessing/comstockpostproc/ami.py:140]: needed to set index col and dtype when loading from csv

File: postprocessing/comstockpostproc/ami.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            ami_name = 'ami_' + lookup_name
+            ami_raw_df = pd.DataFrame()
+            for building_type in self.building_types:
+                file_name = lookup_name + '_' + building_type + '_kwh_per_sqft.csv'
+                file_path = os.path.join(self.truth_data_dir, file_name)
+                if not os.path.isfile(file_path):
+                    logger.warning(f'ami data for ' + file_name + ' does not exist.')
+                    continue
+
+                df = pd.read_csv(file_path, parse_dates=['timestamp'], index_col='timestamp')
```
Comment (eringold) [inline postprocessing/comstockpostproc/ami.py:177]: index col and dtype set when not reloading from csv

File: postprocessing/comstockpostproc/comstock.py [hunk 3]
```diff
@@ -229,6 +237,134 @@ def download_data(self):
...
 
+    def download_timeseries_data_for_ami_comparison(self, ami, reload_from_csv=True, save_individual_regions=False):
+        if reload_from_csv:
+            file_name = f'Timeseries for AMI long.csv'
+            file_path = os.path.join(self.output_dir, file_name)
+            if not os.path.exists(file_path):
+                 raise FileNotFoundError(
+                    f'Cannot find {file_path} to reload data, set reload_from_csv=False to create CSV.')
+            logger.info(f'Reloading from CSV: {file_path}')
+            self.ami_timeseries_data = pd.read_csv(file_path, low_memory=False, index_col='timestamp', parse_dates=True)
```
Comment (eringold) [inline postprocessing/comstockpostproc/comstock.py:249]: set index col and dtype when loading from csv

File: postprocessing/comstockpostproc/comstock.py [hunk 4]
```diff
@@ -229,6 +237,134 @@ def download_data(self):
...
+            "total_site_electricity_kwh": "total"
+        })
+
+        # aggregate by hour
+        agg_df['year'] = agg_df.index.year
+        agg_df['month'] = agg_df.index.month
+        agg_df['day'] = agg_df.index.day
+        agg_df['hour'] = agg_df.index.hour
+        agg_df = agg_df.groupby(['building_type', 'year', 'month', 'day', 'hour']).sum().reset_index()
+        agg_df['timestamp'] = agg_df.apply(
```
Comment (eringold) [inline postprocessing/comstockpostproc/comstock.py:303]: timestamp col as datetime

File: postprocessing/comstockpostproc/comstock.py [hunk 5]
```diff
@@ -229,6 +237,134 @@ def download_data(self):
...
+            agg_df.reset_index(),
+            id_vars=[
+                'timestamp',
+                'building_type',
+                'sample_count'
+            ],
+            value_vars=val_vars,
+            var_name='enduse',
+            value_name='kwh'
+        ).set_index('timestamp')
```
Comment (eringold) [inline postprocessing/comstockpostproc/comstock.py:340]: timestamp as index

**Review by eringold — DISMISSED** [review]
Plots generate when downloading or reloading data. Looks good!

**Comment by eringold** [comment]
@mdahlhausen Please see if everything still works on your end. If so, I think it's good to merge.

**Review by asparke2 — CHANGES_REQUESTED** [review]
**What, if any content should I add to the postprocessing README?**
For the README, write a little section like [this](https://github.com/NREL/ComStock/blob/ami_comparisons/postprocessing/README.md#comparing-one-or-more-comstock-runs-to-each-other-and-cbecs) that explains how to do an AMI comparison. @eringold committed a template script already, so just describe how to use it.

**Existing Plots and Code**
The plots now work for me as well.  Some of them have a slightly clipped peak (see example below), but I don't know that it's important enough to bother fixing unless it's something quick. Otherwise I think the code is fine.

image

**Other Plots**
Two other plots that could be useful:
1. A comparison of all the AMI for each building type plotted on one plot. Could do it normalized and absolute. Basically to give people an understanding of how similar/not the AMI is from region to region.
2. The plot that showed an example week of data for both AMI and ComStock. Sometimes it's easier to pick out things like a time shift when you can see the whole week at once.

**Where to Release Comparisons**
I think we'd like to release the comparisons with the next EUSS release. The first version might just be a folder with a bunch of images on OEDI. We could also try to make a LaTeX template document that just has the hard-coded paths in it and generate a PDF with just a bunch of figures and little/no discussion.

**Comment by mdahlhausen** [comment]
Thanks @eringold and @asparke2.

I'll address two items before merging:
- Adding README content
- The clipped y-axis on some plots

I've been working on summary graphic and statistics that we can use in reports and to guide calibration at a high level. I'll add those in a follow-up PR, along with the two other requested graphs (a plot comparing AMI data, week realizations).

**Comment by mdahlhausen** [comment]
@asparke2:
- reran comparison script from new install of comstock repo and new environment
- added missing truth data
- cast dtypes for polars dataframes
- added README content
- fixed y max issue on day normalized plots
- fixed CBECS comparison test

Should be good to merge now.

I will add the week realizations and a summary of AMI comparisons for the documentation in a future PR

**Review by asparke2 — APPROVED** [review]

**Pull request merged**